yarp/yarp_compiler.c

```diff
@@ -1240,6 +1240,38 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
 
           return;
       }
+      case YP_NODE_SOURCE_ENCODING_NODE: {
+          const char *encoding = compile_context->parser->encoding.name;
+          if (!popped) {
+              rb_encoding *enc = rb_find_encoding(rb_str_new_cstr(encoding));
+              VALUE enc_name = rb_str_new_cstr(enc->name);
+              VALUE enc_str = rb_str_freeze(rb_sprintf("#<Encoding:%"PRIsVALUE">", enc_name));
```

Comment:
I don't think we want the name of the encoding, I think we want the actual encoding object here.

Reply:
Hm, I'm not following what you mean. [The `rb_obj_as_string` call](https://github.com/ruby/ruby/blob/master/sprintf.c#L1128) doesn't work for an `rb_encoding`. Can you say more about what you mean here?

Reply:
I mean the value going into the instruction is the encoding itself, not the name of the encoding. Check `RubyVM::InstructionSequence.compile("__ENCODING__").to_a[13][2]`. It shouldn't be the `rb_encoding`, it should be the Ruby `Encoding` object. You probably need `rb_enc_from_encoding` to convert it.